### ISSUE TITLE
pic: rendering differences with groff

### ISSUE THREAD

**Issue by bapt** [n-t-roff/heirloom-doctools#16]
Using following document: https://svnweb.freebsd.org/base/head/share/doc/papers/bufbio/bio.ms?view=markup

non of the drawings are properly rendered

Command used:
pic bio.ms | nroff -ms

**Comment by n-t-roff** [comment]
Again col -x is missing. This is always required for the original nroff.

pic had intentionally designed for troff only but *not* for nroff. It uses \D drawing commands which had by intention not be implemented in nroff because they are of limited use there.

There are (or had been) plans to implement \D in heirloom nroff but (because of the limited use) this has very low priority. If it could be shown that there are cases for which \D in nroff would be essential this priority would rise. Please use troff for viewing this document (i.e.
```
pic bio.ms | troff -ms | dpost >bio.ps
```
or
```
pic bio.ms | troff -ms | dpost | ps2pdf - bio.pdf
```
(I see there is a font problem in the document. `.ft C` should be `.ft CW` and `\fC` should be `\f(CW`.)

**Comment by bapt** [comment]
Yes I figured out that using troff + dpost (beside the font issue in the document) the graphs where working ok.

My point in the ticket was more on FreeBSD we to provide all those documentations in ascii in /usr/share/doc so anyone can read them with less/more)

I'm trying to see how the heirloom doctools could replace groff in base (and I must say so far it is really promising)

So far the only blocking issues are #12 and this one but if this one is by design (aka pic not designed for that) I need to think of another way probably

**Comment by n-t-roff** [comment]
Implementing -ohang to mdoc(7) is in progress.

pic(1) should in general not be used for documents viewed as ASCII--also not with groff. IMHO the pictures in document you refer are not very informative with GNU nroff.

I think about implementing \D in nroff, so this issue can be kept open. But it may be quite some work and hence may take some time.

In general I suggest to use option -Tlocale (as said) for tbl(1) and nroff to have a more modern rendering based on UTF-8 (in an UTF-8 locale).

**Comment by bapt** [comment]
except that by default freebsd is not using UTF-8 locales and we do ship rendered ascii documents.
Maybe I should rethink the way we are shipping those docs like shipping both a unicode and non unicode version :)

**Comment by n-t-roff** [comment]
`-Tlocale` does use UTF-8 in an UTF-8 locale and ASCII else (equal to `-Tlp`), so it may be a good choice anyway. But that's only a tip, you may use the options you prefer.

**Comment by bapt** [comment]
back to the pic subject, this is probably specific to this document, but actually groff with its ascii output gives usefull drawings

**Comment by n-t-roff** [comment]
I start working on the issue of using pic with nroff now.

**Comment by n-t-roff** [comment]
A very first implementation step of `pic | nroff` is commited now. There are many things which need to be improved or optimized. Please note that pic had never been intented to be used with nroff so nroff is just not prepared for this feature to be implemented. The situation may be different for groff where it could have been an initial design goal to support `pic | nroff`.

For the document in question (bio.ms) the command
```bash
pic bio.ms | nroff -Tlocale -ms | col -x | less
```
in an UTF-8 locale comes somehow "nearest" to groff. There are still many rendering errors, many of them can be fixed.
In an non-UTF-8 locale
```bash
pic bio.ms | nroff -Tlp -ms | col -x | less
```
can be used which does only use simple ASCII characters (`-`, `|`, `+`).
Traditinal lines are drawn with
```bash
pic bio.ms | nroff -ms | col -x | less
```
(For the variants with -T option the option -b needs to be added to col if AT&T col is used becase of a bug in AT&T col. This will be addressed later. FreeBSD col doesn't have this bug.)

For the moment tests had only been done with the document bio.ms and only similar pic drawings are supported currently.

**Comment by bapt** [comment]
Thank you very much ! I all give it a try as soon as I can! That means the only issue I have to figure out to import into freebsd is the sensitive case as stated in #17 because we have a hooks that forbid importing things that could prevent checking out our tree on a non sensitive case filesystem :)

That said thank you very much for all the hard work!

**Comment by n-t-roff** [comment]
As said it is a first implementation step (i.e. an early "RFC").

I now continue with #17 and come back to #16 after #17 is solved (I had not replied to it but I agree that a solution is needed and compatibility to groff would be good even if the heirloom implementation would historically be correct. But compatibility with existing documents will be kept too (new filenames but fonts can be selected under new and old name). I did not analize it yet, when there is new information it will be found under #17.)

**Comment by saper** [comment]
Example:


```
                         V1V2V3

                          | | |
                          | | |
                        __|_|_|
                        | vinum |
                        |       |
             ad0s1e     |_______|        ad3s4f
           ad0s1b                      ad3s4e
         ad0s1a     ad1s1c    ad2s1c ad3s4a

            | | |      |         |      | | |
        |   | | |      |         |      | | |   |
        | __|_|_|  ____|__   ____|__  __|_|_|   |
        |slice/labslice/labeslice/labslice/label|
        | |      | |       | |      | |       | |
        | |______| |_Disk-mini-layer| |_______| |
        |     |        |         |        |     |
              |        |         |        |
          ____|__  ____|__   ____|__  ____|__
         disk (ad0disk (ad1)disk (ad2disk (ad3)
          |      | |       | |      | |       |
          |______| |_______| |______| |_______|



```

@bapt also figured out that this file is broken -stable

**Comment by n-t-roff** [comment]
When rendered in UTF-8 the output of heirloom is:
```
                         V1V2V3

                          │ │ │
                          │ │ │
                        ┌─┴─┴─┘
                        │ vinum │
             ad0s1e     │       │        ad3s4f
           ad0s1b       └───────┘      ad3s4e

         ad0s1a     ad1s1c    ad2s1c ad3s4a
            │ │ │      │         │      │ │ │
            │ │ │      │         │      │ │ │
          ┌─┴─┴─│  ┌───┴──   ┌───┴──┐ ┌─┴─┴─┘
          │      │ │       │ │      │ │       │
         slice/labslice/labeslice/labslice/label
          └───┬──┘ └───┬───┘ └───┬──┘ └───┬───┘
              │      Disk‐mini‐layer      │
              │        │         │        │
          ┌───┴──┐ ┌───┴──   ┌───┴──┐ ┌───┴──
          │      │ │       │ │      │ │       │
         disk (ad0disk (ad1)disk (ad2disk (ad3)
          └──────┘ └───────┘ └──────┘ └───────┘
```
And the output of groff is:
```
                             V1V2V3
                              │ │ │
                              ┴ ┴ ┴
                              │ │ │
                             ┌┴─┴─┴─┐
                             vinum  │
                 ad0s1e      │      │        ad3s4f
                             └─┬──┤─┘
               ad0s1b          ┴           ad3s4e
             ad0s1a │   ad1s1c    ad2s1c ad3s4a │
                ┴ ┴ ┴      ┴         ┴      ┴ ┴ ┴
            ┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼
            ┼ ┌─┴─┴─┴┐  ┌──┴───┐ ┌───┴──┐  ┌┴─┴─┴─┐ ┼
            ┼ │      │  │      │ │      │  │      │ ┼
            slice/labeslice/labslice/labeslice/label┼
            ┼ └───┬──┘  └──┬───┘ └───┬──┘  └──┬───┘ ┼
            ┼┼┼┼┼┼┴┼┼┼┼┼Disk‐mini‐layer┼┼┼┼┼┼┼┴┼┼┼┼┼┼
                  │        │         │        │
              ┌───┴──┐  ┌──┴───┐ ┌───┴──┐  ┌──┴───┐
              │      │  │      │ │      │  │      │
             disk (ad0disk (ad1)disk (ad2disk (ad3)
              └──────┘  └──────┘ └──────┘  └──────┘
```
I know that there are many error. Many of them can be solved but all could only be solved with very much changes to pic and nroff. I think it is not worth it (the latter case). IMHO also the groff terminal output is not very nice, IMHO this file looks best when typeset (troff).

**Comment by n-t-roff** [comment]
What do you mean with "broken -stable"?

**Comment by bapt** [comment]
I think he meant FreeBSD stable :)
@saper note that I'm only aiming at working on this on current :)

**Comment by n-t-roff** [comment]
Ah, ok. Before releases I test 8.4, 9.3 and 10.1 too. But during development it may happen that this or that system doesn't work (but should not happen--sorry for that...).

**Comment by n-t-roff** [comment]
I just tested 10.1 and it compiles and works the same as in current.
@saper please don't use a FreeBSD's col(1) which is older than that in current. bapt did fix col which does now work better than that of heirloom (unfortunately ;-), but I aim to fix heirlooms col eventually too :)

**Comment by bapt** [comment]
@saper note that my col fixes are in stable/10 so will be in 10.2
@n-t-roff heirloom's col and FreeBSD's col are still quite close in term of code, you can just pick up my changes :) btw most of them have also been sync with OpenBSD's

**Comment by n-t-roff** [comment]
From first view it does not look like that they have the same origin. Heirloom's col is from AT&T but FreeBSD's is likely from any xBSD (and very likely not from AT&T). I'll have a look but at the moment it's not important since only the new pic stuff is affected and your col works.

**Comment by saper** [comment]
@bapt I just updated my 10 -stable server yesterday and wanted to have `bio.ascii.gz` as a baseline to compare, but for example page 14 is broken.

@n-t-roff I know it is work in progress, I just wanted to give a quick feedback. Nice picutres in nroff are extremely cool.

**Comment by n-t-roff** [comment]
Except for arrows (which are not drawn currently) this is now solved for nroff without -T option or with option -T37. For *this* document (bio.ms) it is also solved for options -Tlp and -Tlocale (UTF-8). If there are problems with the latter two options with other documents please give me a note.
Arrows will be supported for device -Tlp and -Tlocale in the future.

**Comment by bapt** [comment]
Thanks
